src/pages/signup.rs

```diff
@@ -111,206 +114,237 @@ pub fn Signup(cx: Scope) -> Element {
+        auth.set_username(username.get().clone(), false);
+        auth.set_password(password.get().clone());
 
         cx.spawn({
-            to_owned![logged_in, login_info, username, password, client, error, flows, session];
-
-            info!("{:?}", login_info.read());
+            to_owned![auth, username, password, client, error, flows, session_ref, homeserver];
 
             async move {
-                let response = prepare_register(
-                    login_info.read().homeserver.clone(),
-                    login_info.read().username.clone(),
-                    login_info.read().password.clone(),
-                )
-                .await;
-
-                match response {
-                    Ok(_) => todo!(),
-                    Err(ref error) => match error {
-                        Error::Http(HttpError::UiaaError(FromHttpResponseError::Server(
-                            ServerError::Known(x),
-                        ))) => match x {
-                            uiaa::UiaaResponse::AuthResponse(y) => {
-                                let completed = &y.completed;
-                                let mut flows_to_complete: Vec<AuthType> = vec![];
-
-                                y.flows[0].stages.iter().for_each(|f| {
-                                    if completed.iter().find(|e| *e == f).is_none() {
-                                        flows_to_complete.push(f.clone());
-                                    }
-                                    session.set(y.session.clone());
-
-                                    match f {
-                                        AuthType::ReCaptcha => {
-                                            let x = y.params.deref().get();
-                                            let uiaa_response: Result<
-                                                HashMap<String, _>,
-                                                serde_json::Error,
-                                            > = serde_json::from_str(x);
-
-                                            match uiaa_response {
-                                                Ok(u) => {
-                                                    let m = u.get("m.login.recaptcha");
-
-                                                    if let Some(Value::Object(ref recaptcha)) = m {
-                                                        if let Some(Value::String(public_key)) =
-                                                            recaptcha.get("public_key")
-                                                        {
-                                                            gloo::storage::LocalStorage::set(
-                                                                "sitekey",
-                                                                public_key.clone(),
-                                                            );
+                let login_config = auth.build();
```

Comment:
This  whole block is highly nested, using `let-else` and some helper functions of `Result` could flatten things a bit and make it more readable